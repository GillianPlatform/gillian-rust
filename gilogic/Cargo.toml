[package]
name = "gilogic"
version = "0.1.0"
edition = "2021"

# See more keys and their definitions at https://doc.rust-lang.org/cargo/reference/manifest.html

[dependencies]
gilogic_proc = { path = "../gilogic_proc" }
gilogic_proc_dummy = { path = "../gilogic_proc_dummy" }