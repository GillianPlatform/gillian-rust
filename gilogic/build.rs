fn main() {
    println!("cargo::rustc-check-cfg=cfg(gillian)");
}
