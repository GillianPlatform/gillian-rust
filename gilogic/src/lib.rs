#![allow(internal_features)]
#![feature(rustc_attrs)]
#![feature(never_type)]
#![feature(register_tool)]
#![feature(ptr_internals)]
#![feature(allocator_api)]
#![feature(tuple_trait, unboxed_closures)]
#![feature(stmt_expr_attributes, proc_macro_hygiene)]
#![register_tool(gillian)]

mod tys;

pub use tys::RustAssertion;

pub mod macros {
    pub use gilogic_proc::{
        assertion, assertion_test, borrow, close_borrow, extract_lemma, lemma, open_borrow,
        predicate, show_safety, specification,
    };
    pub mod prophecies {
        pub use gilogic_proc::with_freeze_lemma_for_mutref_pcy as with_freeze_lemma_for_mutref;
    }
    pub mod no_prophecies {
        pub use gilogic_proc::with_freeze_lemma_for_mutref_no_pcy as with_freeze_lemma_for_mutref;
    }
}

mod seq;
pub use seq::Seq;

mod ownable;
pub use ownable::Ownable;
pub mod alloc;
pub mod iterated;
pub mod prophecies;

pub mod symex_ctrl {

    #[gillian::builtin]
    #[gillian::no_translate]
    #[rustc_diagnostic_item = "gillian::symex_ctrl::symex_next"]
    #[inline(always)]
    pub fn symex_next() {}
}

#[path = "stubs.rs"]
pub mod __stubs;
