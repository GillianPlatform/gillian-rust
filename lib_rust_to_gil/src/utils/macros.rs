macro_rules! fatal {
  ($e: expr, $($tts:tt)*) => {
      $e.ty_ctxt.sess.fatal(&format!($($tts)*))
  };
}

pub(crate) use fatal;
