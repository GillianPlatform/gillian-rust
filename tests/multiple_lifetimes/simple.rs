//@check-pass
extern crate gilogic;
use gilogic::macros::*;
use gilogic::Ownable;

// #[show_safety]
fn swap_borrows<'a, 'b>(x : &'a mut i32, y: &'b mut i32) -> (&'b mut i32, &'a mut i32) {
  (y, x)
}

#[show_safety]
fn fst<'a, 'b>(x : &'a mut i32, y: &'b mut i32) -> (&'a mut i32) {
  x
}


#[show_safety]
fn snd<'a, 'b>(x : &'a mut i32, y: &'b mut i32) -> (&'b mut i32) {
  y
}

// #[show_safety]
// fn swap<T: Ownable, U : Ownable>(x : T, y : U) -> (U, T) {
//   (y, x)
// }

// #[show_safety]
// fn swap_borrows2<'a, 'b>(x : &'a mut i32, y: &'b mut i32) -> (&'b mut i32, &'a mut i32) {
//   swap(x, y)
// }

// fn test() {
//   let mut x = 0;
//   let mut y = 1;

//   let (a, b) = swap_borrows2(&mut x, &mut y);
//   let (c, d) = swap_borrows(a, b);

//   *c = 5;
//   *d = 10;

//   assert!(x == 5);
//   assert!(y == 10);
// }