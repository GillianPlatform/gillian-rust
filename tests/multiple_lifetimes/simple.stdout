#begin_init_data
{}
#end_init_data

import "i__binop.gil", "i__lang.gil", "i__std_shims.gil", "i__prophecies.gil", "i__ownable.gil", "i__std_shims_no_pcy.gil";

pred "<&mut i32 as gilogic::Ownable>::own"(+pLft_erased, +self):
  <value>(l-nth(#self, 0i), l-nth(#self, 1i), "i32"; #v)
  * "<i32 as gilogic::Ownable>::own"(#v)
  * types(#lvar_0: Obj, #lvar_1: List) * (#self == {{ #lvar_0, #lvar_1 }})
  * (self == #self);
guard: <lft>(pLft_erased; true);

pred "<&mut i32 as gilogic::Ownable>::own$$inner"(+self):
  <value>(l-nth(#self, 0i), l-nth(#self, 1i), "i32"; #v)
  * "<i32 as gilogic::Ownable>::own"(#v)
  * types(#lvar_0: Obj, #lvar_1: List) * (#self == {{ #lvar_0, #lvar_1 }})
  * (self == #self);


proc swap_borrows(pLft_a, pLft_b, x, y) {
        m_1x := x;
        g_0 := m_1x;
        m_1x := [alloc]({{ "ref", true, "i32" }});
        u := [store_value](l-nth(m_1x, 0i), l-nth(m_1x, 1i), {{ "ref", true, "i32" }}, g_0);
        m_2y := y;
        g_1 := m_2y;
        m_2y := [alloc]({{ "ref", true, "i32" }});
        u := [store_value](l-nth(m_2y, 0i), l-nth(m_2y, 1i), {{ "ref", true, "i32" }}, g_1);
  bb0:  g_3 := [load_value](l-nth(m_2y, 0i), l-nth(m_2y, 1i), {{ "ref", true, "i32" }}, true);
        g_2 := l-nth(g_3, 0i);
        g_4 := [new_lft]();
        pLft_10 := l-nth(g_4, 0i);
        m_3 := g_2;
        g_6 := [load_value](l-nth(m_1x, 0i), l-nth(m_1x, 1i), {{ "ref", true, "i32" }}, true);
        g_5 := l-nth(g_6, 0i);
        g_7 := [new_lft]();
        pLft_11 := l-nth(g_7, 0i);
        m_4 := g_5;
        g_8 := m_3;
        g_9 := m_4;
        m_0 := {{ g_8, g_9 }};
        g_10 := m_0;
        ret := g_10;
        goto rlab;
  rlab: u := [free](l-nth(m_1x, 0i), l-nth(m_1x, 1i), {{ "ref", true, "i32" }});
        u := [free](l-nth(m_2y, 0i), l-nth(m_2y, 1i), {{ "ref", true, "i32" }});
        return
};

spec fst(pLft_a, pLft_b, x, y)
[[ "<&mut i32 as gilogic::Ownable>::own"(#pLft_a, #x) * "<&mut i32 as gilogic::Ownable>::own"(#pLft_a, #y) * <lft>(#pLft_b; true) * <lft>(#pLft_a; true) * types(#lvar_2: Obj, #lvar_3: List) * (#y == {{ #lvar_2, #lvar_3 }}) * types(#lvar_0: Obj, #lvar_1: List) * (#x == {{ #lvar_0, #lvar_1 }}) * (y == #y) * (x == #x) * (pLft_b == #pLft_b) * (pLft_a == #pLft_a) ]]
[["<&mut i32 as gilogic::Ownable>::own"(#pLft_a, ret) * <lft>(#pLft_b; true) * <lft>(#pLft_a; true)]]
normal

proc fst(pLft_a, pLft_b, x, y) {
        m_1x := x;
        g_0 := m_1x;
        m_1x := [alloc]({{ "ref", true, "i32" }});
        u := [store_value](l-nth(m_1x, 0i), l-nth(m_1x, 1i), {{ "ref", true, "i32" }}, g_0);
        m_2y := y;
        m_3 := [alloc]({{ "ref", true, "i32" }});
  bb0:  g_2 := [load_value](l-nth(m_1x, 0i), l-nth(m_1x, 1i), {{ "ref", true, "i32" }}, true);
        g_1 := l-nth(g_2, 0i);
        g_3 := [new_lft]();
        pLft_9 := l-nth(g_3, 0i);
        u := [store_value](l-nth(m_3, 0i), l-nth(m_3, 1i), {{ "ref", true, "i32" }}, g_1);
        g_5 := [load_value](l-nth(m_3, 0i), l-nth(m_3, 1i), {{ "ref", true, "i32" }}, true);
        g_4 := l-nth(g_5, 0i);
        g_6 := [new_lft]();
        pLft_6 := l-nth(g_6, 0i);
        m_0 := g_4;
        g_7 := m_0;
        ret := g_7;
        goto rlab;
  rlab: u := [free](l-nth(m_1x, 0i), l-nth(m_1x, 1i), {{ "ref", true, "i32" }});
        u := [free](l-nth(m_3, 0i), l-nth(m_3, 1i), {{ "ref", true, "i32" }});
        return
};

spec snd(pLft_a, pLft_b, x, y)
[[ "<&mut i32 as gilogic::Ownable>::own"(#pLft_a, #x) * "<&mut i32 as gilogic::Ownable>::own"(#pLft_a, #y) * <lft>(#pLft_b; true) * <lft>(#pLft_a; true) * types(#lvar_2: Obj, #lvar_3: List) * (#y == {{ #lvar_2, #lvar_3 }}) * types(#lvar_0: Obj, #lvar_1: List) * (#x == {{ #lvar_0, #lvar_1 }}) * (y == #y) * (x == #x) * (pLft_b == #pLft_b) * (pLft_a == #pLft_a) ]]
[["<&mut i32 as gilogic::Ownable>::own"(#pLft_a, ret) * <lft>(#pLft_b; true) * <lft>(#pLft_a; true)]]
normal

proc snd(pLft_a, pLft_b, x, y) {
        m_1x := x;
        m_2y := y;
        g_0 := m_2y;
        m_2y := [alloc]({{ "ref", true, "i32" }});
        u := [store_value](l-nth(m_2y, 0i), l-nth(m_2y, 1i), {{ "ref", true, "i32" }}, g_0);
        m_3 := [alloc]({{ "ref", true, "i32" }});
  bb0:  g_2 := [load_value](l-nth(m_2y, 0i), l-nth(m_2y, 1i), {{ "ref", true, "i32" }}, true);
        g_1 := l-nth(g_2, 0i);
        g_3 := [new_lft]();
        pLft_9 := l-nth(g_3, 0i);
        u := [store_value](l-nth(m_3, 0i), l-nth(m_3, 1i), {{ "ref", true, "i32" }}, g_1);
        g_5 := [load_value](l-nth(m_3, 0i), l-nth(m_3, 1i), {{ "ref", true, "i32" }}, true);
        g_4 := l-nth(g_5, 0i);
        g_6 := [new_lft]();
        pLft_6 := l-nth(g_6, 0i);
        m_0 := g_4;
        g_7 := m_0;
        ret := g_7;
        goto rlab;
  rlab: u := [free](l-nth(m_2y, 0i), l-nth(m_2y, 1i), {{ "ref", true, "i32" }});
        u := [free](l-nth(m_3, 0i), l-nth(m_3, 1i), {{ "ref", true, "i32" }});
        return
};

